pipenv/project.py

```diff
@@ -240,12 +240,42 @@ def requirements_exists(self) -> bool:
         return bool(self.requirements_location)
 
     def is_venv_in_project(self) -> bool:
-        if self.s.PIPENV_VENV_IN_PROJECT is False:
-            return False
-        return self.s.PIPENV_VENV_IN_PROJECT or (
-            self.project_directory
-            and os.path.isdir(os.path.join(self.project_directory, ".venv"))
-        )
+        """
+        There are three sources of truth to reconcile:
+
+        - PIPENV_VENV_IN_PROJECT environment variable
+        - venv_in_project directive in Pipfile
+        - an existing .venv directory
+
+        If the Pipfile and ENV var disagree, there are a few options:
+
+        1. Semantic precedence:
+            - if either is True, then VENV_IN_PROJECT=true
+            - if one is False and the other is None, then VENV_IN_PROJECT=false
+        2. Pipfile precedence over ENV Var
+        3. ENV Var precedence over Pipfile
+
+        Whichever one we choose, the Pipfile + Env Var will resolve to true/false/None,
+        and we can then follow the same logic:
```

Comment:
I traced back the origins of the existing logic to a couple of commits:

- https://github.com/pypa/pipenv/commit/949ee95d6748e8777bed589f0d990aa4792b28f8#diff-6dead0c29851e31224fc582f2d7a605a35c3dba132f122fda4f7946736b519ceR1 which explains the explicit "if False return False" clause.
    - I personally think it's a bit clearer to express this as "use whether a current `.venv` directory exists as our truth value *only if no value is set*, otherwise return the explicitly set value".
- https://github.com/justin-yan/pipenv/commit/1e0882c74da5a3074bd5c8f2834f8a151966233f#diff-30286674f55dc3069ed2258ba9962214f350a718c031fb37bd8afda55fef357bR220 which explains why we have `self.project_directory` in an `or` conjunction.
    - I believe that this is no longer necessary - if you look at the code at that point in time, the `project_directory` property was capable of returning `None` which I believe necessitated that check before calling `os.path.join`, whereas now it always returns a string.